spices/SPICE-0013-bytes-standard-library.adoc

```diff
@@ -0,0 +1,376 @@
+= Bytes standard library class
+
+* Proposal: link:./SPICE-0013-bytes-standard-library.adoc[SPICE-0013]
+* Author: https://github.com/bioball[Dan Chao]
+* Status: TBD
+* Implemented in: TBD
+* Category: Standard Library
+
+== Introduction
+
+This SPICE introduces a new value type called `Bytes`, which represents binary data.
+
+== Motivation
+
+Pkl currently does not have a binary data type.
+
+This introduces the following limitations:
+
+* `pkl eval` can only produce UTF-8 encoded strings.
+* Asides from values coming from `read()`, it is not possible to compute checksums of binary content.
+* It is harder to interop with configuration formats that allow binary data.
+
+== Proposed Solution
+
+A new class will be added to `pkl:base`, called `Bytes`.
+
+== Detailed design
+
+=== Standard library changes
+
+The following changes will be made to the standard library
+
+==== `Bytes` class
+
+A new class will be added to `pkl:base`, called `Bytes`.
+This will be a new value type, similar to `Duration` and `DataSize`.
+
+[source,pkl]
+----
+@Since { version = "0.29.0" }
+external class Bytes extends Any {
+  /// The value of these bytes.
+  external value: List<UInt8>
+
+  /// The value represented by these bytes, in base64 encoding.
+  external base64: String(isBase64)
+
+  /// The value represented by these bytes, in hexadecimal encoding.
+  external hex: String
+
+  /// The [MD5](https://en.wikipedia.org/wiki/MD5) hash of these bytes as hexadecimal string.
+  ///
+  /// MD5 is cryptographically broken and should not be used for secure applications.
+  external md5: String
+
+  /// The [SHA-1](https://en.wikipedia.org/wiki/SHA-1) hash of these bytes as hexadecimal string.
+  ///
+  /// SHA-1 is cryptographically broken and should not be used for secure applications.
+  external sha1: String
```

Comment:
I think that the very wide consensus in IT sphere currently is that MD5 and SHA-1 hashes should not be used anymore, so adding them to the new API seems a bit fishy. What is the intended value of having them, especially MD5?

Maybe instead of providing hashes directly on the class level (which is, admittedly, convenient but which is also pretty inflexible) it would make sense to provide a carefully designed module specifically for hashing, which would provide hashing functions instead? It feels like it would be more flexible and consistent, especially if that module will grow with new hash algorithms in future.

Reply:
> Maybe instead of providing hashes directly on the class level (which is, admittedly, convenient but which is also pretty inflexible) it would make sense to provide a carefully designed module specifically for hashing, which would provide hashing functions instead? It feels like it would be more flexible and consistent, especially if that module will grow with new hash algorithms in future.

We can maybe add a `pkl:crypto` module or something similar in the future, if the need arises. But, these methods are also included because we're migrating the existing `Resource.{sha1,sha256,etc}` methods into `Bytes`.